pkg/querier/merge_queryable.go

```diff
@@ -0,0 +1,280 @@
+	}
+	labelNames, warnings, err := mergeDistinctStringSlice(funcs...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check if the tenant label exists in the original result
+	var tenantLabelExists bool
+	for _, labelName := range labelNames {
+		if labelName == string(m.tenantLabelName) {
+			tenantLabelExists = true
+			break
+		}
+	}
+
+	if tenantLabelExists {
+		// the tenant label exists in the original result, so it had been retained
+		labelNames = append(labelNames, retainExistingPrefix+string(m.tenantLabelName))
+	} else {
+		// the tenant label was not existing, so it need to get added
+		labelNames = append(labelNames, string(m.tenantLabelName))
+	}
+	sort.Strings(labelNames)
+
+	return labelNames, warnings, nil
+
+}
+
+// Close releases the resources of the Querier.
+func (m *mergeQuerier) Close() error {
+	var errs prometheus.MultiError
+	for pos := range m.tenantIDs {
+		if err := m.queriers[pos].Close(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to close querier for %s: %w", m.tenantIDs, err))
+		}
+	}
+	return errs.MaybeUnwrap()
+}
+
+// Select returns a set of series that matches the given label matchers.
+// Caller can specify if it requires returned series to be sorted. Prefer not requiring sorting for better performance.
+// It allows passing hints that can help in optimising select, but it's up to implementation how this is used if used at all.
+func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+	tenantIDsPos, filteredMatchers := filterValuesByMatchers(string(m.tenantLabelName), m.tenantIDs, matchers...)
+	var seriesSets = make([]storage.SeriesSet, len(tenantIDsPos))
+	for pos, posTenant := range tenantIDsPos {
+		tenantID := m.tenantIDs[posTenant]
+		seriesSets[pos] = &addLabelsSeriesSet{
+			upstream: m.queriers[posTenant].Select(sortSeries, hints, filteredMatchers...),
+			labels: labels.Labels{
+				{
+					Name:  string(m.tenantLabelName),
+					Value: tenantID,
+				},
+			},
+		}
+	}
+	return storage.NewMergeSeriesSet(seriesSets, storage.ChainedSeriesMerge)
+}
+
+type addLabelsSeriesSet struct {
+	upstream storage.SeriesSet
+	labels   labels.Labels
+}
+
+func (m *addLabelsSeriesSet) Next() bool {
+	return m.upstream.Next()
+}
+
+// At returns full series. Returned series should be iteratable even after Next is called.
+func (m *addLabelsSeriesSet) At() storage.Series {
+	return &addLabelsSeries{
+		upstream: m.upstream.At(),
+		labels:   m.labels,
+	}
+}
+
+// The error that iteration as failed with.
+// When an error occurs, set cannot continue to iterate.
+func (m *addLabelsSeriesSet) Err() error {
+	return m.upstream.Err()
+}
+
+// A collection of warnings for the whole set.
+// Warnings could be return even iteration has not failed with error.
+func (m *addLabelsSeriesSet) Warnings() storage.Warnings {
+	return m.upstream.Warnings()
+}
+
+type addLabelsSeries struct {
+	upstream storage.Series
+	labels   labels.Labels
```

Comment:
**question:** why make this multiple labels, we only need one to store the tenant?

Reply:
I thought about having potentially multiple dimensions in the future (e.g. for federation across clusters)

Reply:
I see, not sure if we will be able to reuse it. However, the flexibility of this implementation doesn't hurt either.